google-cloud-pubsub/lib/google/cloud/pubsub/topic/async_publisher.rb

```diff
@@ -0,0 +1,323 @@
+        #
+        #   topic = pubsub.topic "my-topic"
+        #   topic.publish_async "task completed" do |result|
+        #     if result.succeeded?
+        #       log_publish_success result.data
+        #     else
+        #       log_publish_failure result.data, result.error
+        #     end
+        #   end
+        #
+        #   topic.async_publisher.stop.wait!
+        #
+        # @attr_reader [String] topic_name The name of the topic the messages
+        #   are published to. In the form of
+        #   "/projects/project-identifier/topics/topic-name".
+        # @attr_reader [Integer] max_bytes The maximum size of messages to be
+        #   collected before the batch is published. Default is 10,000,000
+        #   (10MB).
+        # @attr_reader [Integer] max_messages The maximum number of messages to
+        #   be collected before the batch is published. Default is 1,000.
+        # @attr_reader [Numeric] interval The number of seconds to collect
+        #   messages before the batch is published. Default is 0.25.
+        # @attr_reader [Numeric] publish_threads The number of threads used to
+        #   publish messages. Default is 4.
+        # @attr_reader [Numeric] callback_threads The number of threads to
+        #   handle the published messages' callbacks. Default is 8.
+        #
+        class AsyncPublisher
+          include MonitorMixin
+
+          attr_reader :topic_name, :max_bytes, :max_messages, :interval,
+                      :publish_threads, :callback_threads
+          ##
+          # @private Implementation accessors
+          attr_reader :service, :batch, :publish_thread_pool,
+                      :callback_thread_pool
+
+          ##
+          # @private Create a new instance of the object.
+          def initialize topic_name, service, max_bytes: 10000000,
+                         max_messages: 1000, interval: 0.25, threads: {}
+            @topic_name = service.topic_path topic_name
+            @service    = service
+
+            @max_bytes        = max_bytes
+            @max_messages     = max_messages
+            @interval         = interval
+            @publish_threads  = (threads[:publish] || 4).to_i
+            @callback_threads = (threads[:callback] || 8).to_i
+
+            @cond = new_cond
+
+            # init MonitorMixin
+            super()
+          end
+
+          ##
+          # Add a message to the async publisher to be published to the topic.
+          # Messages will be collected in batches and published together.
+          # See {Google::Cloud::Pubsub::Topic#publish_async}
+          def publish data = nil, attributes = {}, &block
+            msg = create_pubsub_message data, attributes
+
+            synchronize do
+              fail "Can't publish when stopped." if @stopped
+
+              if @batch.nil?
+                @batch ||= Batch.new self
+                @batch.add msg, block
+              else
+                unless @batch.try_add msg, block
```

Comment:
We should consider messages that are by themselves larger than the batching limits. i.e. max batched bytes is 10 and we publish a 100 byte message. We would here be adding it to the batch without publish_batch being called immediately. I think it were better that if the message was larger than the batch we published the batch and it by itself immediately.